lib/remote/jsonrpcconnection-pki.cpp

```diff
@@ -289,6 +269,20 @@ void JsonRpcConnection::SendCertificateRequest(const JsonRpcConnection::Ptr& acl
 
 	/* Path is empty if this is our own request. */
 	if (path.IsEmpty()) {
+		if (aclient && !origin) {
+			Log(LogInformation, "JsonRpcConnection")
+				<< "Requesting new certificate for this Icinga instance from endpoint '" << aclient->GetIdentity() << "'.";
+
+			if (!aclient->m_SendCsrTimer) {
+				aclient->m_SendCsrTimer = new Timer();
+				aclient->m_SendCsrTimer->OnTimerExpired.connect([aclient](const Timer * const&) {
+					aclient->SendCertificateRequest(aclient, nullptr, "");
+				});
+				aclient->m_SendCsrTimer->SetInterval(29 * 24 * 60 * 60);
+				aclient->m_SendCsrTimer->Start();
+			}
+		}
```

Comment:
I think this code should only become active when called from https://github.com/Icinga/icinga2/blob/ce4d816d40f17c0123e8f7196ad5185264123ccf/lib/remote/apilistener.cpp#L837

Why not simply start the timer from there instead (possibly using some `aclient->StartRenewalTimer()` helper function).

Reply:
Do you prefer two methods instead of one?

The idea is: oh, this conn. may renew our certs once = this conn. may renew our certs once again.

Reply:
Compare this code with there:
https://github.com/Icinga/icinga2/blob/352a89b575c7071cdc83d40627b29b822076afb2/lib/remote/apilistener.cpp#L239-L246

There you renew the certificate once and then start the timer doing it regularly. I thought you just put the code here because otherwise you'd have to access members of `JsonRpcConnection` in `ApiListener`.

Also starting the timer from
https://github.com/Icinga/icinga2/blob/ce4d816d40f17c0123e8f7196ad5185264123ccf/lib/remote/apilistener.cpp#L837
wouldn't require these additional checks in here, so simpler code.

Reply:
`aclient && !origin`? Still needed for the log message.

Reply:
Consider this comment thread my attempt of trying to keep this function more readable. I think this function is already somewhat strange in general (sometimes sends no certificate, making the remote implicitly use the certificate used for the TLS connection, sometime it sends one from a file, sometimes reading back in a file that was just written right before the function was called). And this really doesn't get any better if it additionally starts a timer now - but only under some conditions guarded by three layers of ifs.

It's not like the log message isn't something that can be solved either. It could be logged within the lambda instead for example. And no, that wouldn't require duplicating the log message, that lambda could be called synchronously once before starting the timer, or maybe, given that the response is asynchronous I think, simply rescheduling the timer to fire immediately should work as well.

Reply:
I don’t exactly understand how you see the log message deduplicated.

Anyway: in general I consider the function fine. It does its job.